[AArch64] Add a FeatureFuseCmpCSet

davemgreen opened the pull request:
This adds a new feature, FeatureFuseCmpCSet, and adds it to cpus that according to the SWOG can fuse cmp+cset. The existing FeatureFuseCCSelect was renamed FeatureFuseCmpCSel for consistency.

SamTebbs33 commented on llvm/lib/Target/AArch64/AArch64MacroFusion.cpp:
```diff
@@ -279,6 +279,40 @@ static bool isCCSelectPair(const MachineInstr *FirstMI,
+static bool isCmpCSetPair(const MachineInstr *FirstMI,
+                          const MachineInstr &SecondMI) {
+  if ((SecondMI.getOpcode() == AArch64::CSINCWr &&
+       SecondMI.getOperand(1).getReg() == AArch64::WZR &&
+       SecondMI.getOperand(2).getReg() == AArch64::WZR) ||
+      (SecondMI.getOpcode() == AArch64::CSINCXr &&
+       SecondMI.getOperand(1).getReg() == AArch64::XZR &&
+       SecondMI.getOperand(2).getReg() == AArch64::XZR)) {
```
Ah I see, thanks.

SamTebbs33 reviewed (approved):
Looks good to me, thanks.

rj-jesus commented on llvm/lib/Target/AArch64/AArch64MacroFusion.cpp:
```diff
@@ -279,6 +279,40 @@ static bool isCCSelectPair(const MachineInstr *FirstMI,
+    if (FirstMI == nullptr)
+      return true;
+
+    if (FirstMI->definesRegister(AArch64::WZR, /*TRI=*/nullptr) ||
+        FirstMI->definesRegister(AArch64::XZR, /*TRI=*/nullptr))
```
Should this be broken down per 32-bit/64-bit variant, as done in `isCmpCSelPair`? (I'm not sure if that matters.)

davemgreen committed:
[AArch64] Add a FeatureFuseCmpCSet

This adds a new feature, FeatureFuseCmpCSet, and adds it to cpus that according
to the SWOG can fuse cmp+cset. The existing FeatureFuseCCSelect was renamed
FeatureFuseCmpCSel for consistency
